### ISSUE TITLE
async/await vs returning promise

### ISSUE THREAD

**Issue by arnaudbenard** [avajs/eslint-plugin-ava#162]
Hi,

I have noticed that the linter sometimes recommends me to write the test with `async/await` and sometimes to return a promise. I'm a bit confused on which one to use when and how the linter decides which one to use.

```js
// Return a promise works
test('fails', t => {
  return doSomething({})
    .catch(err => t.truthy(err));
});

// ava/prefer-async-await recommended (return promise does not work)
test('succeeds', async t => {
  const result = await doSomething({ data: {}});
  t.truthy(result);
});
```

Thank you! 👍

**Comment by jfmengels** [comment]
AVA's maintainers will tell you to always use async/await if possible, just because that makes the test shorter and easier to read. It is just a code style preference that some may not adhere to.

The part about returning a Promise is to avoid errors. AVA needs to know when the tests end to make sure your tests are run properly. Tests created with `test.cb()` end when `t.end()` are called. Tests that return a Promise end when that Promise is resolved or rejected (async functions are implicitly transformed into Promises and returned), and tests that return something else or nothing end when the implementation function returns.

In other words, if you have a test that uses promises and does not return anything (explicitly, or implicitly via async/await) will be ended before some async part will be evaluated.

```js
test('should fail but does not', t => {
  doSomething({})
    .then(() => t.fail());
});
```

Here, the test is considered finished before the function passed to `then` is called. That is why it is important to return your Promises.
But as said before, async/await removes that need and makes that simpler :)

Hope that answers your question :)

**Comment by arnaudbenard** [comment]
Awesome answer, thank you very much for the feedback 👍
